### PR TITLE
Presto statistics

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+package com.facebook.presto.spi.statistics;
+
+import java.util.Optional;
+
+public class ColumnStatistics
+{
+    public static final ColumnStatistics EMPTY_STATISTICS = ColumnStatistics.builder().build();
+
+    private final Optional<StatisticsValue<Long>> dataSize;
```
Comment (kbajda) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Is this a size of a single column value? Like 8 bytes for BIGINT?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    private final Optional<StatisticsValue<Long>> rowsCount;
```
Comment (kbajda) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Wouldn't that be one value for entire table rather than per column stat?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    private final Optional<StatisticsValue<Object>> min;
```
Comment (kbajda) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: maybe minValue?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    private final Optional<StatisticsValue<Object>> max;
```
Comment (kbajda) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: maxValue? also, how about nullCount?

**Comment by kbajda** [comment]
@losipiuk a quick review done
I suppose we are not planning NDV at this point, correct?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    private final boolean fullyReliable;
+
+    private final Optional<Double> reliabilityLevel;
+
+    /**
+     * Statistics reliability in [0,1] range.
+     */
+    public static Reliability reliabilityLevel(Double reliabilityLevel)
+    {
+        if (reliabilityLevel < 0.0 || reliabilityLevel > 1.0) {
```
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: `checkArgument`?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        if (!fullyReliable ^ reliabilityLevel.isPresent()) {
+            throw new IllegalArgumentException("cannot set both fullyReliable to true and specify reliabilityLevel");
+        }
+        this.fullyReliable = fullyReliable;
+        this.reliabilityLevel = reliabilityLevel;
+    }
+
+    public boolean isFullyReliable()
+    {
+        return fullyReliable;
```
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: Shouldn't it be a function of reliability-level instead of bool flag?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    public Double getReliabilityLevel()
+    {
+        if (!reliabilityLevel.isPresent()) {
+            throw new IllegalStateException("reliabilityLevel not set");
```
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: Why not `unknown` special value (or `Double.NaN` as a reliability-level) similarly to how we indicate empty/lack of statistics with `EMPTY_STATISTICS` instance? I would change it here to special values or https://github.com/Teradata/presto/pull/230/files#diff-db7a547329ed26ca7452b9c95b842b84R21 for consistency.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+public class StatisticsValue<T>
+{
+    private final Reliability reliability;
+    private final T value;
+
+    public StatisticsValue(Reliability reliability, T value)
+    {
+        this.reliability = requireNonNull(reliability, "reliability can not be null");
+        this.value = value;
```
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java:35]: Do we really want to accept nulls here?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 2 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Yeah - probably (or rather partition). Having it here (even if it is the same value for all the columns) saves one domain object but is misleading. 
I will move it to new `TableStatistics` class which will be part of `ConnectorTableLayout`.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 1 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: I was rather thinking of size of all values in column. So for BIGINT column it would be `number_of_rows*8`.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java (hunk 5 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: no guava here

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java (hunk 6 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: It is function of reliability-level. I.e it is method on `Reliability` class. Boolean flag is just internal implementation.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java (hunk 7 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: Do you mean that we should expose to users that `Double.NaN` means that there is no reliability? Or just internally treat it as special value to not have `Optional`. I do not like the former - but latter seem ok (though use of `Optional` still seems cleaner to me).

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java (hunk 8 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java:35]: Well - not really. Empty tables and stats like `minValue`/`maxValue` are somewhat tricky though. Any suggestion to handle that?

File: presto-spi/src/main/java/com/facebook/presto/spi/ColumnMetadata.java [hunk 9]
```diff
@@ -13,18 +13,22 @@
...
+import static java.util.Objects.requireNonNull;
 
 public class ColumnMetadata
 {
     private final String name;
     private final Type type;
     private final String comment;
     private final boolean hidden;
+    // todo should columnStatistics be part of toString/equals/hashCode
+    private final ColumnStatistics columnStatistics;
```
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/ColumnMetadata.java]: Second thoughts on this one.
Actually I think this field should be removed from here.
And instead `Map<ColumnHandle,ColumnStatistics>` should be added to `ConnectorTableLayout`.
This way we will be able to support different statistics for different set of table partitions in use.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 4 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: I started with just random few stats. Of course it will grow over time.
I have to check what hive supports out of the box and expose that for a start.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java (hunk 7 above)
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: Having `Double.NaN` enables all arithmetics, is comparable etc. I'd expose it to the user.
Just to make it clear for the user, I'd also expose a method `isReliabilityEstablished` or something like that in addition to returning here `NaN`.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java (hunk 6 above)
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: I understand this. I claim it would be better to avoid boolean flag (which is misleading and error-pron).
E.g.

``` java
public boolean isFullyReliable()
{
  return reliabilityLevel == Double.POSITIVE_INFINITY;
}
```

is much more consistent and less error-pron because you don't have to pass the boolean flag around. FullyReliabilitiness is derived from the actual reliability measure.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java (hunk 8 above)
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java:35]: It's handled on higher level (by having `StatisticsValue` optional here: https://github.com/Teradata/presto/pull/230/files/621e0f3401cec83bb5dcfc1cb9404aee2b214c59#diff-db7a547329ed26ca7452b9c95b842b84R26), so we should ensure it's non-null here.
While lacking value we should make it `Optional.empty()` where we know what min/max means (in `ColumnStatistics`).

**Comment by maciejgrzybek** [comment]
General comment: shouldn't we consider making `ColumnStatistics` JSON-serializable? It could be helpful to make it send-able in e.g. planner.
If yes, we should not be using `Optionals`.
I'm not saying we should annotate it with `@JSONCreator` etc. now but we should at least avoid using `Optional`.

File: presto-spi/src/main/java/com/facebook/presto/spi/ColumnMetadata.java (hunk 9 above)
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/ColumnMetadata.java]: +1

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 4 above)
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Myabe `min` and `max` should be grouped together. How one can have `min` without having `max`?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+ * limitations under the License.
+ */
+
+package com.facebook.presto.spi.statistics;
+
+import static java.util.Objects.requireNonNull;
+
+public class StatisticsValue<T>
+{
+    private final Reliability reliability;
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java]: should not `reliability` be put at the table level?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java (hunk 8 above)
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java:35]: For empty tables table level stats should return 0 as row count and the there should be no values at all (no min nor max)

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+package com.facebook.presto.spi.statistics;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class Reliability
+{
+    private final boolean fullyReliable;
```
Comment (kokosing) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: double values `0.0` and `1.0` should have exact representation in double. So maybe instead of having this field you could `reliabilityLevel` set to `1.0`. Having another variable causes some issues, for example when someone creates `Reliability` with value `1.0` then this flag will be set to `false` while stats are fully reliable. 

Also don't you need to have `fullUnreliable` field as well?

**Comment by kokosing** [comment]
IMO stats should be very optional for conectors. In case connector does not support them there should be no need to create any object with `Optional.empty()` values in it. Maybe there should be some optional `StatsProvider` which would return `ColumnStats` or `TableStats` for given `ColumnHandle` or `TableHandle`?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java (hunk 6 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: Honestly - it does not matter at all - as this is hidden inside `100 lines` short class. And is part of internal impl. But I will change to make you happy :)

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java (hunk 7 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: I will leave it in range [0,1] for now + `POSITIVE_INFINITY` as marker for "fully-reliable" (but expose the marker only through `isFullyReliable()` method).
And see if that needs extension based on use cases.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java (hunk 8 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java:35]: I dropped min/max for now. We will resume discussion when we have usecases for that.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java (hunk 10 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java]: As we discussed if it should be moved somewhere it would be column (not table) level.
As we may have different level of knowledge for different columns.
Also let's keep it here for now. And see if it is painful. Theoretically we could have different level of reliability for different stats (e.g. min/max vs ndv)

**Comment by losipiuk** [comment]
> IMO stats should be very optional for conectors. In case connector does not support them there should be no need to create any object with Optional.empty() values in it. Maybe there should be some optional StatsProvider which would return ColumnStats or TableStats for given ColumnHandle or TableHandle?

I do not get it here. What is the problem of adding `EMPTY_STATISTICS` as last parameter in `ConnectorTableLayout`.

**Comment by losipiuk** [comment]
Updated version. Multiple commits but treat those as one. It will be squashed before push.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java [hunk 12]
```diff
@@ -947,6 +948,7 @@ private void rollbackPartitionUpdates(String user, List<PartitionUpdate> partiti
 
     /**
      * Attempts to remove the file or empty directory.
+     *
```
Comment (sopel39) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java]: unrelated changes

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+package com.facebook.presto.spi.statistics;
+
+import java.util.Optional;
+
+public final class ColumnStatistics
+{
+    public static final ColumnStatistics EMPTY_STATISTICS = ColumnStatistics.builder().build();
+
+    private final Optional<StatisticsValue<Long>> dataSize;
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: `Optional` here is redundant and will complicate computations and usage of statistics. How is `Optional.empty()` different (from computations point of view) than `dataSize` with reliability `0`.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.spi.statistics;
+
+public final class Reliability
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: What exactly is `Reliability`? Please add some definition (formal would be best). If it doesn't have formal definition, then it doesn't have much use. If connectors can return arbitrary values, how optimizer knows what is reliable and what isn't? How it can compare reliabilities from different sources?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 13 above)
Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: We are kinda stalled here with @sopel39. Other reviewers - please add comments.

It seems to me that it is somewhat reasonable to get rid of `Optional` wrapper and then just have `reliability` in `SV` set to `0.0` but then what should we use as `SV.value`? Maybe we should change `SV` to only accept `double` and then use `NaN`? What do you think?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java (hunk 14 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/Reliability.java]: Additionally, how can reliability be propagated up?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.spi.statistics;
+
+import static java.util.Objects.requireNonNull;
+
+public final class StatisticsValue<T>
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java]: Binding value with reliability requires us to write reliability algebra (e.g.: `StatisticsValue_a + StatisticsValue_b = StatisticsValue_c, *, /,` etc). Should we care about it? Can't we just have one reliability for entire column or table?

Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java]: It will not be used through whole system so painfulness of operating on `SV` should not be that bad.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 13 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Another alternative would be to have `reliability` (aka `fractionData`) as separate field. All stats values could be `doubles` with `NaN` when stats are missing. Direct `double` values are easier to work with. Additionally, `NaNs` propagate nicely so formulas are compact (e.g `a + b` instead of `statisicsValue(dataFraction(a.dataFraction.value + b.dataFraction.value), a.value + b.value)`, not to mention more complex formulas)

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/DataFraction.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.spi.statistics;
+
+public final class DataFraction
+{
+    private static final double FULLY_RELIABLE = Double.POSITIVE_INFINITY;
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/DataFraction.java]: `FULLY_RELIABLE -> ALL_DATA`?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/TableStatistics.java [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public final class TableStatistics
+{
+    public static final TableStatistics EMPTY_STATISTICS = TableStatistics.builder().build();
+
+    private final Optional<StatisticsValue<Long>> rowsCount;
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/TableStatistics.java]: `Long` -> `Double`? Use `NaN` for missing stat?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/TableStatistics.java [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+package com.facebook.presto.spi.statistics;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public final class TableStatistics
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/TableStatistics.java:24]: Maybe just have `fractionData` field here and in `ColumnStatistics` instead of attaching it to each individual value? It would make API cleaner and computations easier. What @maciejgrzybek and @kokosing  think?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 4 above)
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: I don't see value in grouping them but there could be some value in providing partial stats.
E.g. scenario: connector is assuming DB is used mainly through Presto so stats are refreshed on inserts/updates through Presto but from time to time they are refreshed from remote system (in case there are outside Presto DMLs).
So refresh time = last refresh / DML + TTL
Assuming `min` value was established recently by insert through Presto but no one updated `max` value using Presto for a while. Then stats for `min` value can be provided but unless connector refreshes stats from remote system `max` can't be provided.

It may not be a real case but shows there's some explanation for such an approach. And given the fact there's no benefit of merging them together I'd keep them separate.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java (hunk 15 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java]: Still, what when we add `min, max, ndv` and `ranges` stats? Will they also have `StatisticsValue` attached? Can't there be just one `fractionRead` field per column? Also, won't `ColumnStatistics` be used to aggregate stats above scan node? Will there be other class for that?

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 4 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: In your scenario, how `min` can be established but `max` doesn't. The point is: maybe we shouldnt overengineer for now and focus on simplicity and clear formulas. In reality min will come with max most of the times.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/TableStatistics.java (hunk 17 above)
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/TableStatistics.java]: Can you have a fraction number of rows?
We have `Optional` for missing stats.

Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/TableStatistics.java]: Yes, but with `NaN` you can just write `a + b` and `NaN` will propagate. With optionals you have to write:

```
if (a.isPresent() && b.isPresent()) {
  return Optional.of(a.get() + b.get());
} else {
  return Optional.empty();
}
```

for more complex formulas it is even more cumbersome

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java (hunk 4 above)
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/ColumnStatistics.java]: Prerequisites:
empty table, no stats provided

Steps:
1. insert into x values(2); // min updated, max updated
2. insert into x values(5); // max updated
3. // wait TTL (to invalidate max stats)
4. insert into x values(1); // min updated

Then min is updated by DML but max is not updated unless refresh is triggered (because of TTL).

It's not overengineering. I just claim to leave it as is instead of introducing any fancy mechanism (that would be overengineering). I don't see any value of merging min and max and gave some scenario as a proof there might be many of similar scenarios we don't think of now. And argument "let's merge and see later" is not an argument.
So if there is at least one argument against merging and no argument for merging... Answer is simple.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/TableStatistics.java (hunk 17 above)
Comment (maciejgrzybek) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/TableStatistics.java]: So this is exactly what I was suggesting previously. ;)
https://github.com/Teradata/presto/pull/230/files/621e0f3401cec83bb5dcfc1cb9404aee2b214c59#r68155946

**Comment by losipiuk** [comment]
Ok - we have some ideas for proper approach after discussion with @fiedukow and @sopel39. I will update the PR tomorrow. Thanks guys.

**Comment by losipiuk** [comment]
Ok guys - this is simplified. Take a look now.

Summary of changes:
- `reliability` was changed to `estimatorVariance` which has an understandable meaning.
- `StatisticsValue` is no longer parametrized with type and always expose `double` value. This does not play well if we want to expose values from column domain (like `min` or `max`) as `SV`. But simplifies other cases. And `min`/`max` should rather be boundaries of histograms for which different statistics are computed as `SV` - and not `SV` itself.
- I dropped `Optional` wrapper around `StatisticsValue`. Unknown statistic is exposed as `Double.NaN` (helper methods/constant are also available). This will simplify computations on `SV`.

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java [hunk 19]
```diff
@@ -14,16 +14,24 @@
...
 
-import static java.util.Objects.requireNonNull;
-
-public final class StatisticsValue<T>
+public final class StatisticsValue
 {
-    private final DataFraction dataFraction;
-    private final T value;
+    private final double value;
+    private final double estimatorVariance;
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java]: Unfortunately, `estimatorVariance` cannot be computed when using sample data (e.g. statistics) to compute population mean (e.g average row length). This is because variance for the entire population is unknown. In such case, T-student distribution might be used to estimate mean with given confidence level. Please have a read:
https://en.wikipedia.org/wiki/Student%27s_t-distribution#Sampling_distribution
https://en.wikipedia.org/wiki/Confidence_interval#Practical_example
https://en.wikipedia.org/wiki/Confidence_interval#Theoretical_example
https://en.wikipedia.org/wiki/Variance#Sample_variance
 
I suggest using confidence intervals to measure estimate "accuracy".

File: presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+package com.facebook.presto.spi.statistics;
+
+public final class StatisticsValue
+{
+    public static final double UNKNOWN_STATISTICS = Double.NaN;
+
+    private final double value;
+    private final double estimatorVariance;
```
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java]: Unfortunately, estimatorVariance cannot be computed when using sample data to compute population mean (e.g average row length). This is because variance for the entire population is unknown. In such case, T-student distribution might be used to estimate mean with given confidence level. Please have a read:
https://en.wikipedia.org/wiki/Student%27s_t-distribution#Sampling_distribution
https://en.wikipedia.org/wiki/Confidence_interval#Practical_example
https://en.wikipedia.org/wiki/Confidence_interval#Theoretical_example
https://en.wikipedia.org/wiki/Variance#Sample_variance

Therefore I suggest using confidence intervals to measure estimate "accuracy".

Comment (losipiuk) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java]: Or maybe just drop this accuracy alltogether for now?

Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/statistics/StatisticsValue.java]: Sure. However, we could describe FB our proposal for "accuracy" to check what is their opinion.

**Comment by losipiuk** [comment]
I removed the `estimatorVariance` for now.
New review with code to read stats from HiveConnector is here: https://github.com/Teradata/presto/pull/239
Closing this one
